File: src/agent.rs
// Built-in agent definitions and detection.
//
// Each AgentDef carries two install paths: `project_path` (relative, used when
// installing into a repo) and `global_path` (joined under $HOME, used for
// `--global`). `detect_dir` is the per-user probe used by auto-detection;
// it's empty for target-only agents (e.g. "universal") that the user must
// opt into with `-a`.
//
// `aliases` holds deprecated/alternate names. Looking up an agent by alias
// works but emits a warning so the user migrates to the canonical name.

use std::path::PathBuf;

#[derive(Debug, Clone, Copy)]
pub struct AgentDef {
    pub name: &'static str,
    pub display: &'static str,
    pub aliases: &'static [&'static str],
    pub project_path: &'static str,
    pub global_path: &'static str,
    pub detect_dir: &'static str,
    pub binary: Option<&'static str>,
}

#[rustfmt::skip]
pub const AGENT_DEFS: &[AgentDef] = &[
    // --- Original 12 (paths fixed for amplify->augment, aider->.aider-desk, copilot) ---
    AgentDef { name: "claude",     display: "Claude Code",     aliases: &[],           project_path: ".claude/skills",     global_path: ".claude/skills",          detect_dir: ".claude",          binary: Some("claude") },
    AgentDef { name: "cursor",     display: "Cursor",          aliases: &[],           project_path: ".cursor/skills",     global_path: ".cursor/skills",          detect_dir: ".cursor",          binary: Some("cursor") },
    AgentDef { name: "opencode",   display: "OpenCode",        aliases: &[],           project_path: ".opencode/skills",   global_path: ".config/opencode/skills", detect_dir: ".config/opencode", binary: Some("opencode") },
    AgentDef { name: "cline",      display: "Cline",           aliases: &[],           project_path: ".cline/skills",      global_path: ".cline/skills",           detect_dir: ".cline",           binary: None },
    AgentDef { name: "codex",      display: "Codex",           aliases: &[],           project_path: ".codex/skills",      global_path: ".codex/skills",           detect_dir: ".codex",           binary: Some("codex") },
    AgentDef { name: "windsurf",   display: "Windsurf",        aliases: &[],           project_path: ".windsurf/skills",   global_path: ".codeium/windsurf/skills", detect_dir: ".windsurf",       binary: None },
    AgentDef { name: "continue",   display: "Continue",        aliases: &[],           project_path: ".continue/skills",   global_path: ".continue/skills",        detect_dir: ".continue",        binary: None },
    AgentDef { name: "copilot",    display: "GitHub Copilot",  aliases: &[],           project_path: ".agents/skills",     global_path: ".copilot/skills",         detect_dir: ".copilot",         binary: None },
    AgentDef { name: "aider",      display: "AiderDesk",       aliases: &[],           project_path: ".aider-desk/skills", global_path: ".aider-desk/skills",      detect_dir: ".aider-desk",      binary: None },
    AgentDef { name: "roo",        display: "Roo",             aliases: &[],           project_path: ".roo/skills",        global_path: ".roo/skills",             detect_dir: ".roo",             binary: None },
    AgentDef { name: "augment",    display: "Augment Code",    aliases: &["amplify"],  project_path: ".augment/skills",    global_path: ".augment/skills",         detect_dir: ".augment",         binary: None },
    AgentDef { name: "zed",        display: "Zed",             aliases: &[],           project_path: ".zed/skills",        global_path: ".zed/skills",             detect_dir: ".zed",             binary: Some("zed") },

    // --- Tier 1 additions (verified against the agent's own docs) ---
    AgentDef { name: "gemini-cli", display: "Gemini CLI",      aliases: &[],           project_path: ".agents/skills",     global_path: ".gemini/skills",          detect_dir: ".gemini",          binary: Some("gemini") },
    AgentDef { name: "goose",      display: "Goose",           aliases: &[],           project_path: ".goose/skills",      global_path: ".config/goose/skills",    detect_dir: ".config/goose",    binary: Some("goose") },
    AgentDef { name: "kilo",       display: "Kilo Code",       aliases: &[],           project_path: ".kilocode/skills",   global_path: ".kilocode/skills",        detect_dir: ".kilocode",        binary: None },
    AgentDef { name: "warp",       display: "Warp",            aliases: &[],           project_path: ".agents/skills",     global_path: ".agents/skills",          detect_dir: ".warp",            binary: Some("warp") },
    AgentDef { name: "amp",        display: "Amp",             aliases: &[],           project_path: ".agents/skills",     global_path: ".config/agents/skills",   detect_dir: ".config/agents",   binary: Some("amp") },
    AgentDef { name: "qwen-code",  display: "Qwen Code",       aliases: &[],           project_path: ".qwen/skills",       global_path: ".qwen/skills",            detect_dir: ".qwen",            binary: None },
    AgentDef { name: "crush",      display: "Crush",           aliases: &[],           project_path: ".crush/skills",      global_path: ".config/crush/skills",    detect_dir: ".config/crush",    binary: Some("crush") },
    AgentDef { name: "openhands",  display: "OpenHands",       aliases: &[],           project_path: ".openhands/skills",  global_path: ".openhands/skills",       detect_dir: ".openhands",       binary: None },
    AgentDef { name: "kiro-cli",   display: "Kiro CLI",        aliases: &[],           project_path: ".kiro/skills",       global_path: ".kiro/skills",            detect_dir: ".kiro",            binary: None },
    AgentDef { name: "tabnine-cli", display: "Tabnine CLI",    aliases: &[],           project_path: ".tabnine/agent/skills", global_path: ".tabnine/agent/skills", detect_dir: ".tabnine",       binary: None },

    // --- Tier 2/3 additions (paths from skills npm README; not all independently verified) ---
    AgentDef { name: "aider-desk",     display: "AiderDesk",       aliases: &[],   project_path: ".aider-desk/skills",     global_path: ".aider-desk/skills",            detect_dir: ".aider-desk",         binary: None },
    AgentDef { name: "antigravity",    display: "Antigravity",     aliases: &[],   project_path: ".agents/skills",         global_path: ".gemini/antigravity/skills",    detect_dir: ".gemini/antigravity", binary: None },
    AgentDef { name: "bob",            display: "IBM Bob",         aliases: &[],   project_path: ".bob/skills",            global_path: ".bob/skills",                   detect_dir: ".bob",                binary: None },
    AgentDef { name: "openclaw",       display: "OpenClaw",        aliases: &[],   project_path: "skills",                 global_path: ".openclaw/skills",              detect_dir: ".openclaw",           binary: None },
    AgentDef { name: "codearts-agent", display: "CodeArts Agent",  aliases: &[],   project_path: ".codeartsdoer/skills",   global_path: ".codeartsdoer/skills",          detect_dir: ".codeartsdoer",       binary: None },
    AgentDef { name: "codebuddy",      display: "CodeBuddy",       aliases: &[],   project_path: ".codebuddy/skills",      global_path: ".codebuddy/skills",             detect_dir: ".codebuddy",          binary: None },
    AgentDef { name: "codemaker",      display: "Codemaker",       aliases: &[],   project_path: ".codemaker/skills",      global_path: ".codemaker/skills",             detect_dir: ".codemaker",          binary: None },
    AgentDef { name: "codestudio",     display: "Code Studio",     aliases: &[],   project_path: ".codestudio/skills",     global_path: ".codestudio/skills",            detect_dir: ".codestudio",         binary: None },
    AgentDef { name: "command-code",   display: "Command Code",    aliases: &[],   project_path: ".commandcode/skills",    global_path: ".commandcode/skills",           detect_dir: ".commandcode",        binary: None },
    AgentDef { name: "cortex",         display: "Cortex Code",     aliases: &[],   project_path: ".cortex/skills",         global_path: ".snowflake/cortex/skills",      detect_dir: ".cortex",             binary: None },
    AgentDef { name: "deepagents",     display: "Deep Agents",     aliases: &[],   project_path: ".agents/skills",         global_path: ".deepagents/agent/skills",      detect_dir: ".deepagents",         binary: None },
    AgentDef { name: "devin",          display: "Devin",           aliases: &[],   project_path: ".devin/skills",          global_path: ".config/devin/skills",          detect_dir: ".devin",              binary: None },
    AgentDef { name: "dexto",          display: "Dexto",           aliases: &[],   project_path: ".agents/skills",         global_path: ".agents/skills",                detect_dir: ".dexto",              binary: None },
    AgentDef { name: "droid",          display: "Droid (Factory)", aliases: &[],   project_path: ".factory/skills",        global_path: ".factory/skills",               detect_dir: ".factory",            binary: None },
    AgentDef { name: "firebender",     display: "Firebender",      aliases: &[],   project_path: ".agents/skills",         global_path: ".firebender/skills",            detect_dir: ".firebender",         binary: None },
    AgentDef { name: "forgecode",      display: "ForgeCode",       aliases: &[],   project_path: ".forge/skills",          global_path: ".forge/skills",                 detect_dir: ".forge",              binary: None },
    AgentDef { name: "hermes-agent",   display: "Hermes Agent",    aliases: &[],   project_path: ".hermes/skills",         global_path: ".hermes/skills",                detect_dir: ".hermes",             binary: None },
    AgentDef { name: "iflow-cli",      display: "iFlow CLI",       aliases: &[],   project_path: ".iflow/skills",          global_path: ".iflow/skills",                 detect_dir: ".iflow",              binary: None },
    AgentDef { name: "junie",          display: "Junie",           aliases: &[],   project_path: ".junie/skills",          global_path: ".junie/skills",                 detect_dir: ".junie",              binary: None },
    AgentDef { name: "kimi-cli",       display: "Kimi Code CLI",   aliases: &[],   project_path: ".agents/skills",         global_path: ".config/agents/skills",         detect_dir: ".kimi",               binary: None },
    AgentDef { name: "kode",           display: "Kode",            aliases: &[],   project_path: ".kode/skills",           global_path: ".kode/skills",                  detect_dir: ".kode",               binary: None },
    AgentDef { name: "mcpjam",         display: "MCPJam",          aliases: &[],   project_path: ".mcpjam/skills",         global_path: ".mcpjam/skills",                detect_dir: ".mcpjam",             binary: None },
    AgentDef { name: "mistral-vibe",   display: "Mistral Vibe",    aliases: &[],   project_path: ".vibe/skills",           global_path: ".vibe/skills",                  detect_dir: ".vibe",               binary: None },
    AgentDef { name: "mux",            display: "Mux",             aliases: &[],   project_path: ".mux/skills",            global_path: ".mux/skills",                   detect_dir: ".mux",                binary: None },
    AgentDef { name: "neovate",        display: "Neovate",         aliases: &[],   project_path: ".neovate/skills",        global_path: ".neovate/skills",               detect_dir: ".neovate",            binary: None },
    AgentDef { name: "pi",             display: "Pi",              aliases: &[],   project_path: ".pi/skills",             global_path: ".pi/agent/skills",              detect_dir: ".pi",                 binary: None },
    AgentDef { name: "pochi",          display: "Pochi",           aliases: &[],   project_path: ".pochi/skills",          global_path: ".pochi/skills",                 detect_dir: ".pochi",              binary: None },
    AgentDef { name: "qoder",          display: "Qoder",           aliases: &[],   project_path: ".qoder/skills",          global_path: ".qoder/skills",                 detect_dir: ".qoder",              binary: None },
    AgentDef { name: "replit",         display: "Replit",          aliases: &[],   project_path: ".agents/skills",         global_path: ".config/agents/skills",         detect_dir: ".replit",             binary: None },
    AgentDef { name: "rovodev",        display: "Rovo Dev",        aliases: &[],   project_path: ".rovodev/skills",        global_path: ".rovodev/skills",               detect_dir: ".rovodev",            binary: None },
    AgentDef { name: "trae",           display: "Trae",            aliases: &[],   project_path: ".trae/skills",           global_path: ".trae/skills",                  detect_dir: ".trae",               binary: None },
    AgentDef { name: "trae-cn",        display: "Trae CN",         aliases: &[],   project_path: ".trae/skills",           global_path: ".trae-cn/skills",               detect_dir: ".trae-cn",            binary: None },
    AgentDef { name: "zencoder",       display: "Zencoder",        aliases: &[],   project_path: ".zencoder/skills",       global_path: ".zencoder/skills",              detect_dir: ".zencoder",           binary: None },
    AgentDef { name: "adal",           display: "AdaL",            aliases: &[],   project_path: ".adal/skills",           global_path: ".adal/skills",                  detect_dir: ".adal",               binary: None },
    AgentDef { name: "universal",      display: "Universal",       aliases: &[],   project_path: ".agents/skills",         global_path: ".config/agents/skills",         detect_dir: "",                    binary: None },
];

#[derive(Debug, Clone)]
pub struct Agent {
    pub def: AgentDef,
    /// Where rosie will install/symlink skills for this agent. Relative
    /// (e.g. ".claude/skills") for local installs, absolute (joined under
    /// $HOME) for global installs.
    pub install_path: PathBuf,
    /// True if `$HOME/<detect_dir>` exists or the user passed `-a <name>`.
    pub detected: bool,
}

/// Look up an agent by name or alias. When matching by alias, emit a
/// deprecation warning so users migrate to the canonical name.
pub fn find_agent_def(name: &str) -> Option<&'static AgentDef> {
    if let Some(d) = AGENT_DEFS.iter().find(|d| d.name == name) {
        return Some(d);
    }
    if let Some(d) = AGENT_DEFS.iter().find(|d| d.aliases.contains(&name)) {
        crate::log::warn(&format!(
            "agent name '{}' is deprecated; use '{}'",
            name, d.name
        ));
        return Some(d);
    }
    None
}

/// Where rosie will write/symlink skills for a given agent.
pub fn install_path(def: &AgentDef, global: bool) -> Option<PathBuf> {
    if global {
        let home = crate::os::home_dir()?;
        Some(PathBuf::from(home).join(def.global_path))
    } else {
        Some(PathBuf::from(def.project_path))
    }
}

/// Detect all agents by checking `$HOME/<detect_dir>`. Agents with an empty
/// `detect_dir` are skipped — those are target-only (use `-a <name>`).
pub fn detect_agents(global: bool) -> Vec<Agent> {
    let mut out = Vec::new();
    let home = match crate::os::home_dir() {
        Some(h) => h,
        None => {
            crate::log::error("Cannot determine home directory");
            return out;
        }
    };

    for def in AGENT_DEFS {
        if def.detect_dir.is_empty() {
            continue;
        }
        let probe = PathBuf::from(&home).join(def.detect_dir);
        if crate::os::is_dir(&probe) {
            if let Some(path) = install_path(def, global) {
                crate::log::debug(&format!(
                    "Detected agent: {} ({})",
                    def.display,
                    probe.display()
                ));
                out.push(Agent {
                    def: *def,
                    install_path: path,
                    detected: true,
                });
            }
        }
    }

    out
}

/// Build an AgentList from explicit names (mirrors `-a foo -a bar`).
pub fn agents_from_names(names: &[&str], global: bool) -> Vec<Agent> {
    let mut out = Vec::new();
    for &name in names {
        match find_agent_def(name) {
            Some(def) => {
                if let Some(path) = install_path(def, global) {
                    out.push(Agent {
                        def: *def,
                        install_path: path,
                        detected: true,
                    });
                }
            }
            None => crate::log::error(&format!("Unknown agent: {name}")),
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_count() {
        assert_eq!(AGENT_DEFS.len(), 57);
    }

    #[test]
    fn lookup() {
        assert!(find_agent_def("claude").is_some());
        assert!(find_agent_def("notathing").is_none());
    }

    #[test]
    fn alias_resolves() {
        let d = find_agent_def("amplify").expect("amplify alias should resolve");
        assert_eq!(d.name, "augment");
    }

    #[test]
    fn unique_names() {
        let mut names: Vec<&str> = AGENT_DEFS.iter().map(|d| d.name).collect();
        names.sort();
        let len = names.len();
        names.dedup();
        assert_eq!(names.len(), len, "duplicate agent names in AGENT_DEFS");
    }
}

File: npm/rosie-skills/src/agent.ts
// Built-in agent definitions and detection. Ported from src/agent.rs.
//
// Each AgentDef carries two install paths: projectPath (relative, used when
// installing into a repo) and globalPath (joined under $HOME, for --global).
// detectDir is the per-user probe used by auto-detection; empty for
// target-only agents (e.g. "universal") that need an explicit -a.

import * as path from "node:path";
import * as os from "./os.js";
import * as log from "./log.js";

export interface AgentDef {
  name: string;
  display: string;
  aliases: string[];
  projectPath: string;
  globalPath: string;
  detectDir: string;
  binary: string | null;
}

export const AGENT_DEFS: AgentDef[] = [
  // --- Original 12 ---
  { name: "claude", display: "Claude Code", aliases: [], projectPath: ".claude/skills", globalPath: ".claude/skills", detectDir: ".claude", binary: "claude" },
  { name: "cursor", display: "Cursor", aliases: [], projectPath: ".cursor/skills", globalPath: ".cursor/skills", detectDir: ".cursor", binary: "cursor" },
  { name: "opencode", display: "OpenCode", aliases: [], projectPath: ".opencode/skills", globalPath: ".config/opencode/skills", detectDir: ".config/opencode", binary: "opencode" },
  { name: "cline", display: "Cline", aliases: [], projectPath: ".cline/skills", globalPath: ".cline/skills", detectDir: ".cline", binary: null },
  { name: "codex", display: "Codex", aliases: [], projectPath: ".codex/skills", globalPath: ".codex/skills", detectDir: ".codex", binary: "codex" },
  { name: "windsurf", display: "Windsurf", aliases: [], projectPath: ".windsurf/skills", globalPath: ".codeium/windsurf/skills", detectDir: ".windsurf", binary: null },
  { name: "continue", display: "Continue", aliases: [], projectPath: ".continue/skills", globalPath: ".continue/skills", detectDir: ".continue", binary: null },
  { name: "copilot", display: "GitHub Copilot", aliases: [], projectPath: ".agents/skills", globalPath: ".copilot/skills", detectDir: ".copilot", binary: null },
  { name: "aider", display: "AiderDesk", aliases: [], projectPath: ".aider-desk/skills", globalPath: ".aider-desk/skills", detectDir: ".aider-desk", binary: null },
  { name: "roo", display: "Roo", aliases: [], projectPath: ".roo/skills", globalPath: ".roo/skills", detectDir: ".roo", binary: null },
  { name: "augment", display: "Augment Code", aliases: ["amplify"], projectPath: ".augment/skills", globalPath: ".augment/skills", detectDir: ".augment", binary: null },
  { name: "zed", display: "Zed", aliases: [], projectPath: ".zed/skills", globalPath: ".zed/skills", detectDir: ".zed", binary: "zed" },

  // --- Tier 1 additions ---
  { name: "gemini-cli", display: "Gemini CLI", aliases: [], projectPath: ".agents/skills", globalPath: ".gemini/skills", detectDir: ".gemini", binary: "gemini" },
  { name: "goose", display: "Goose", aliases: [], projectPath: ".goose/skills", globalPath: ".config/goose/skills", detectDir: ".config/goose", binary: "goose" },
  { name: "kilo", display: "Kilo Code", aliases: [], projectPath: ".kilocode/skills", globalPath: ".kilocode/skills", detectDir: ".kilocode", binary: null },
  { name: "warp", display: "Warp", aliases: [], projectPath: ".agents/skills", globalPath: ".agents/skills", detectDir: ".warp", binary: "warp" },
  { name: "amp", display: "Amp", aliases: [], projectPath: ".agents/skills", globalPath: ".config/agents/skills", detectDir: ".config/agents", binary: "amp" },
  { name: "qwen-code", display: "Qwen Code", aliases: [], projectPath: ".qwen/skills", globalPath: ".qwen/skills", detectDir: ".qwen", binary: null },
  { name: "crush", display: "Crush", aliases: [], projectPath: ".crush/skills", globalPath: ".config/crush/skills", detectDir: ".config/crush", binary: "crush" },
  { name: "openhands", display: "OpenHands", aliases: [], projectPath: ".openhands/skills", globalPath: ".openhands/skills", detectDir: ".openhands", binary: null },
  { name: "kiro-cli", display: "Kiro CLI", aliases: [], projectPath: ".kiro/skills", globalPath: ".kiro/skills", detectDir: ".kiro", binary: null },
  { name: "tabnine-cli", display: "Tabnine CLI", aliases: [], projectPath: ".tabnine/agent/skills", globalPath: ".tabnine/agent/skills", detectDir: ".tabnine", binary: null },

  // --- Tier 2/3 additions ---
  { name: "aider-desk", display: "AiderDesk", aliases: [], projectPath: ".aider-desk/skills", globalPath: ".aider-desk/skills", detectDir: ".aider-desk", binary: null },
  { name: "antigravity", display: "Antigravity", aliases: [], projectPath: ".agents/skills", globalPath: ".gemini/antigravity/skills", detectDir: ".gemini/antigravity", binary: null },
  { name: "bob", display: "IBM Bob", aliases: [], projectPath: ".bob/skills", globalPath: ".bob/skills", detectDir: ".bob", binary: null },
  { name: "openclaw", display: "OpenClaw", aliases: [], projectPath: "skills", globalPath: ".openclaw/skills", detectDir: ".openclaw", binary: null },
  { name: "codearts-agent", display: "CodeArts Agent", aliases: [], projectPath: ".codeartsdoer/skills", globalPath: ".codeartsdoer/skills", detectDir: ".codeartsdoer", binary: null },
  { name: "codebuddy", display: "CodeBuddy", aliases: [], projectPath: ".codebuddy/skills", globalPath: ".codebuddy/skills", detectDir: ".codebuddy", binary: null },
  { name: "codemaker", display: "Codemaker", aliases: [], projectPath: ".codemaker/skills", globalPath: ".codemaker/skills", detectDir: ".codemaker", binary: null },
  { name: "codestudio", display: "Code Studio", aliases: [], projectPath: ".codestudio/skills", globalPath: ".codestudio/skills", detectDir: ".codestudio", binary: null },
  { name: "command-code", display: "Command Code", aliases: [], projectPath: ".commandcode/skills", globalPath: ".commandcode/skills", detectDir: ".commandcode", binary: null },
  { name: "cortex", display: "Cortex Code", aliases: [], projectPath: ".cortex/skills", globalPath: ".snowflake/cortex/skills", detectDir: ".cortex", binary: null },
  { name: "deepagents", display: "Deep Agents", aliases: [], projectPath: ".agents/skills", globalPath: ".deepagents/agent/skills", detectDir: ".deepagents", binary: null },
  { name: "devin", display: "Devin", aliases: [], projectPath: ".devin/skills", globalPath: ".config/devin/skills", detectDir: ".devin", binary: null },
  { name: "dexto", display: "Dexto", aliases: [], projectPath: ".agents/skills", globalPath: ".agents/skills", detectDir: ".dexto", binary: null },
  { name: "droid", display: "Droid (Factory)", aliases: [], projectPath: ".factory/skills", globalPath: ".factory/skills", detectDir: ".factory", binary: null },
  { name: "firebender", display: "Firebender", aliases: [], projectPath: ".agents/skills", globalPath: ".firebender/skills", detectDir: ".firebender", binary: null },
  { name: "forgecode", display: "ForgeCode", aliases: [], projectPath: ".forge/skills", globalPath: ".forge/skills", detectDir: ".forge", binary: null },
  { name: "hermes-agent", display: "Hermes Agent", aliases: [], projectPath: ".hermes/skills", globalPath: ".hermes/skills", detectDir: ".hermes", binary: null },
  { name: "iflow-cli", display: "iFlow CLI", aliases: [], projectPath: ".iflow/skills", globalPath: ".iflow/skills", detectDir: ".iflow", binary: null },
  { name: "junie", display: "Junie", aliases: [], projectPath: ".junie/skills", globalPath: ".junie/skills", detectDir: ".junie", binary: null },
  { name: "kimi-cli", display: "Kimi Code CLI", aliases: [], projectPath: ".agents/skills", globalPath: ".config/agents/skills", detectDir: ".kimi", binary: null },
  { name: "kode", display: "Kode", aliases: [], projectPath: ".kode/skills", globalPath: ".kode/skills", detectDir: ".kode", binary: null },
  { name: "mcpjam", display: "MCPJam", aliases: [], projectPath: ".mcpjam/skills", globalPath: ".mcpjam/skills", detectDir: ".mcpjam", binary: null },
  { name: "mistral-vibe", display: "Mistral Vibe", aliases: [], projectPath: ".vibe/skills", globalPath: ".vibe/skills", detectDir: ".vibe", binary: null },
  { name: "mux", display: "Mux", aliases: [], projectPath: ".mux/skills", globalPath: ".mux/skills", detectDir: ".mux", binary: null },
  { name: "neovate", display: "Neovate", aliases: [], projectPath: ".neovate/skills", globalPath: ".neovate/skills", detectDir: ".neovate", binary: null },
  { name: "pi", display: "Pi", aliases: [], projectPath: ".pi/skills", globalPath: ".pi/agent/skills", detectDir: ".pi", binary: null },
  { name: "pochi", display: "Pochi", aliases: [], projectPath: ".pochi/skills", globalPath: ".pochi/skills", detectDir: ".pochi", binary: null },
  { name: "qoder", display: "Qoder", aliases: [], projectPath: ".qoder/skills", globalPath: ".qoder/skills", detectDir: ".qoder", binary: null },
  { name: "replit", display: "Replit", aliases: [], projectPath: ".agents/skills", globalPath: ".config/agents/skills", detectDir: ".replit", binary: null },
  { name: "rovodev", display: "Rovo Dev", aliases: [], projectPath: ".rovodev/skills", globalPath: ".rovodev/skills", detectDir: ".rovodev", binary: null },
  { name: "trae", display: "Trae", aliases: [], projectPath: ".trae/skills", globalPath: ".trae/skills", detectDir: ".trae", binary: null },
  { name: "trae-cn", display: "Trae CN", aliases: [], projectPath: ".trae/skills", globalPath: ".trae-cn/skills", detectDir: ".trae-cn", binary: null },
  { name: "zencoder", display: "Zencoder", aliases: [], projectPath: ".zencoder/skills", globalPath: ".zencoder/skills", detectDir: ".zencoder", binary: null },
  { name: "adal", display: "AdaL", aliases: [], projectPath: ".adal/skills", globalPath: ".adal/skills", detectDir: ".adal", binary: null },
  { name: "universal", display: "Universal", aliases: [], projectPath: ".agents/skills", globalPath: ".config/agents/skills", detectDir: "", binary: null },
];

export interface Agent {
  def: AgentDef;
  // Where rosie will install/symlink skills for this agent. Relative (e.g.
  // ".claude/skills") for local installs, absolute (under $HOME) for global.
  installPath: string;
  // True if `$HOME/<detectDir>` exists or the user passed `-a <name>`.
  detected: boolean;
}

// Look up an agent by name or alias. Matching by alias emits a deprecation
// warning so users migrate to the canonical name.
export function findAgentDef(name: string): AgentDef | undefined {
  const direct = AGENT_DEFS.find((d) => d.name === name);
  if (direct) return direct;
  const aliased = AGENT_DEFS.find((d) => d.aliases.includes(name));
  if (aliased) {
    log.warn(`agent name '${name}' is deprecated; use '${aliased.name}'`);
    return aliased;
  }
  return undefined;
}

// Where rosie will write/symlink skills for a given agent.
export function installPath(def: AgentDef, global: boolean): string | null {
  if (global) {
    const home = os.homeDir();
    if (home === null) return null;
    return path.join(home, def.globalPath);
  }
  return def.projectPath;
}

// Detect all agents by checking `$HOME/<detectDir>`. Agents with an empty
// detectDir are skipped (target-only; use -a <name>).
export function detectAgents(global: boolean): Agent[] {
  const out: Agent[] = [];
  const home = os.homeDir();
  if (home === null) {
    log.error("Cannot determine home directory");
    return out;
  }
  for (const def of AGENT_DEFS) {
    if (def.detectDir.length === 0) continue;
    const probe = path.join(home, def.detectDir);
    if (os.isDir(probe)) {
      const ip = installPath(def, global);
      if (ip !== null) {
        log.debug(`Detected agent: ${def.display} (${probe})`);
        out.push({ def, installPath: ip, detected: true });
      }
    }
  }
  return out;
}

// Build an agent list from explicit names (mirrors `-a foo -a bar`).
export function agentsFromNames(names: string[], global: boolean): Agent[] {
  const out: Agent[] = [];
  for (const name of names) {
    const def = findAgentDef(name);
    if (def) {
      const ip = installPath(def, global);
      if (ip !== null) out.push({ def, installPath: ip, detected: true });
    } else {
      log.error(`Unknown agent: ${name}`);
    }
  }
  return out;
}
